File: Instruens Fabulam/fab.rb
# Function to parse input and return a list of tables
def parse_input(input_lines)
    tables = []
    current_table = []
  
    input_lines.each do |line|
      line.strip!
      if line == '*'
        tables << current_table unless current_table.empty?
        break
      elsif line.chars.all? { |ch| ['<', '=', '>'].include?(ch) }
        tables << current_table unless current_table.empty?
        current_table = [line.chars]
      else
        current_table << line.split('&').map(&:strip)
      end
    end
  
    tables
  end
  
  # Function to calculate the maximum width for each column
  def calculate_widths(table)
    table.transpose.map do |column|
      column.map(&:length).max
    end
  end
  
  # Function to format the table according to the specification
  def format_table(table)
    alignments = table[0]  # The first row contains the alignment markers
    widths = calculate_widths(table[1..])  # Calculate the widths excluding the alignment row
  
    # Define alignment mappings for left, center, and right
    align_map = { '<' => :ljust, '=' => :center, '>' => :rjust }
  
    # Format the header row
    header = '| ' + table[1].zip(alignments, widths).map do |text, align, width|
      text.strip.send(align_map[align], width)
    end.join(' | ') + ' |'
  
    # Create the top and bottom borders with appropriate lengths
    divider = '|' + widths.map { |width| '-' * (width + 2) }.join('+') + '|'
    top_border = '@' + '-' * (divider.length - 2) + '@'
  
    # Format each row of the table (after the header)
    rows = table[2..].map do |row|
      '| ' + row.zip(alignments, widths).map do |text, align, width|
        text.strip.send(align_map[align], width)
      end.join(' | ') + ' |'
    end
  
    # Return the formatted table as a string
    [top_border, header, divider, *rows, top_border].join("\n")
  end
  
  # Main function to read input, format tables, and output results
  def main
    # Read input file
    input_lines = File.readlines('fab.in').map(&:chomp)
  
    # Parse input
    tables = parse_input(input_lines)
  
    # Open output file for writing
    File.open('fab.out', 'w') do |output_file|
      tables.each_with_index do |table, i|
        formatted_table = format_table(table)
  
        # Print the formatted table without an automatic newline
        print formatted_table
        output_file.write(formatted_table)
  
        # Manually add a newline only between tables, but not after the last one
        if i < tables.size - 1
          print "\n"  # Newline for the console
          output_file.write("\n")  # Newline for the file
        end
      end
    end
  end
    
  # Call the main function
  main if __FILE__ == $PROGRAM_NAME

File: Instruens Fabulam/fab.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func parseInput(inputLines []string) [][][]string {
	tables := [][][]string{}
	currentTable := [][]string{}

	for _, line := range inputLines {
		line = strings.TrimSpace(line)
		if line == "*" {
			if len(currentTable) > 0 {
				tables = append(tables, currentTable)
			}
			break
		} else if isAlignmentLine(line) {
			if len(currentTable) > 0 {
				tables = append(tables, currentTable)
			}
			currentTable = [][]string{strings.Split(line, "")}
		} else {
			cells := strings.Split(line, "&")
			for i := range cells {
				cells[i] = strings.TrimSpace(cells[i])
			}
			currentTable = append(currentTable, cells)
		}
	}

	return tables
}

func isAlignmentLine(line string) bool {
	for _, char := range line {
		if char != '<' && char != '=' && char != '>' {
			return false
		}
	}
	return true
}

func calculateWidths(table [][]string) []int {
	widths := make([]int, len(table[0]))
	for i := range widths {
		maxWidth := 0
		for _, row := range table[1:] {
			if len(row[i]) > maxWidth {
				maxWidth = len(row[i])
			}
		}
		widths[i] = maxWidth
	}
	return widths
}

func formatTable(table [][]string) []string {
	alignments := table[0]
	widths := calculateWidths(table)
	formattedRows := []string{}

	for _, row := range table[1:] {
		formattedCells := []string{}
		for i, cell := range row {
			width := widths[i]
			var formattedCell string

			switch alignments[i] {
			case "<":
				formattedCell = fmt.Sprintf("%-*s", width, cell)
			case "=":
				padding := (width - len(cell)) / 2
				formattedCell = fmt.Sprintf("%*s%s%*s", padding, "", cell, width-padding-len(cell), "")
			case ">":
				formattedCell = fmt.Sprintf("%*s", width, cell)
			}
			formattedCells = append(formattedCells, formattedCell)
		}
		formattedRow := "| " + strings.Join(formattedCells, " | ") + " |"
		formattedRows = append(formattedRows, formattedRow)
	}

	topBorder := "@" + strings.Repeat("-", len(formattedRows[0])-2) + "@"
	divider := "|" + strings.Join(makeDivider(widths), "+") + "|"

	formattedTable := []string{topBorder, formattedRows[0], divider}
	formattedTable = append(formattedTable, formattedRows[1:]...)
	formattedTable = append(formattedTable, topBorder)
	return formattedTable
}

func makeDivider(widths []int) []string {
	dividers := []string{}
	for _, width := range widths {
		dividers = append(dividers, strings.Repeat("-", width+2))
	}
	return dividers
}

func main() {
	file, err := os.Open("fab.in")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var inputLines []string
	for scanner.Scan() {
		inputLines = append(inputLines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	outputFile, err := os.Create("fab.out")
	if err != nil {
		fmt.Println("Error creating output file:", err)
		return
	}
	defer outputFile.Close()

	writer := bufio.NewWriter(outputFile)
	defer writer.Flush()

	tables := parseInput(inputLines)
	for i, table := range tables {
		formattedTable := formatTable(table)
		for j, line := range formattedTable {
			if j > 0 || i > 0 {
				fmt.Fprintln(writer) // Add newline only between tables
			}
			fmt.Fprint(writer, line) // Prevent extra trailing newline
		}
	}
}
